pkg/helpers/apply.go

```diff
@@ -0,0 +1,177 @@
+// Copyright Contributors to the Open Cluster Management project
+package helpers
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+)
+
+var (
+	genericScheme = runtime.NewScheme()
+	genericCodecs = serializer.NewCodecFactory(genericScheme)
+	genericCodec  = genericCodecs.UniversalDeserializer()
+)
+
+func ApplyDeployment(
+	client kubernetes.Interface,
+	reader ScenarioReader,
+	templateName string,
+	values interface{},
+	files ...string) error {
+	genericScheme.AddKnownTypes(appsv1.SchemeGroupVersion, &appsv1.Deployment{})
+	recorder := events.NewInMemoryRecorder(GetExampleHeader())
+	for _, name := range files {
+		deploymentBytes, err := mustTempalteAsset(name, templateName, reader, values)
+		if err != nil {
+			return err
+		}
+		deployment, sch, err := genericCodec.Decode(deploymentBytes, nil, nil)
+		if err != nil {
+			return fmt.Errorf("%q: %v %v", name, sch, err)
+		}
+		_, _, err = resourceapply.ApplyDeployment(
+			client.AppsV1(),
+			recorder,
+			deployment.(*appsv1.Deployment), 0)
+		if err != nil {
+			return fmt.Errorf("%q (%T): %v", name, deployment, err)
+		}
+	}
+	return nil
+}
+
+func ApplyDirectly(clients *resourceapply.ClientHolder,
+	reader ScenarioReader,
+	templateName string,
+	values interface{},
+	files ...string) error {
+	recorder := events.NewInMemoryRecorder(GetExampleHeader())
+	resourceResults := resourceapply.ApplyDirectly(clients, recorder, func(name string) ([]byte, error) {
+		return mustTempalteAsset(name, templateName, reader, values)
+	}, files...)
+	for _, result := range resourceResults {
+		if result.Error != nil {
+			return fmt.Errorf("%q (%T): %v", result.File, result.Type, result.Error)
+		}
+	}
+	return nil
+}
+
+func ApplyCustomResouces(client dynamic.Interface,
+	reader ScenarioReader,
+	templateName string,
+	values interface{},
+	files ...string) error {
+	// recorder := events.NewInMemoryRecorder(GetExampleHeader())
+	for _, name := range files {
+		asset, err := mustTempalteAsset(name, templateName, reader, values)
+		if err != nil {
+			return err
+		}
+		u, err := bytesToUnstructured(reader, asset)
+		if err != nil {
+			return err
+		}
+		gvks, _, err := genericScheme.ObjectKinds(u)
+		if err != nil {
+			return err
+		}
+		gvk := gvks[0]
+		resource, err := getResource(u.GetKind())
+		if err != nil {
+			return err
+		}
+		dr := client.Resource(gvk.GroupVersion().WithResource(resource))
+		if u.GetNamespace() != "" {
```

Comment:
you do not need this if. client will not add ns in the path if there is none

Reply:
Done